### PR TITLE
Introduce an object with combined state

### PR DESCRIPTION
**Pull request by ljedrz**
This PR is purely a refactoring and despite the rather large diff, it introduces no changes to the current functionalities, other than the order of some operations when the node is started.

There are 2 main changes:
1. [rename `Status` to `RawStatus` and `State` to `Status`](https://github.com/AleoHQ/snarkOS/commit/0851bd0c10456a35234ad92e6d4d7b8ceda194db): both of these objects refer to the same thing, so their names should match as well; also, this allows the new combined state object to be named `State`
2. [introduce a `State` object](https://github.com/AleoHQ/snarkOS/commit/0ee6be33ec34c154b56fb5c112610950b5a9d70e): the current node structure involves a few actor-like objects that communicate with one another; this is fine, but the state of the actors overlaps, can't really be fully isolated, and adds plenty of extra complexity; this change proposes a combined `State` object shared between the main actors, allowing the communication between them to be much simpler, while not changing the way their channels are utilized or the scopes of their functionalities; this setup reduces code complexity and will enable further improvements to the overall structure

### CODE REVIEW

**Review by niklaslong — COMMENTED** [review]
I'll rebase this onto the branch with the RTT metrics and run it for a bit—should give us a perf sanity check.

**Review by niklaslong — APPROVED** [review]
The RTT metrics would suggest the performance is unchanged—I'll leave the branch running for a while longer.

I think this is a desirable first step towards consolidating state. The need for sync primitives should dissapear once the channels are consolidated. 👍

File: network/src/peer.rs [hunk 1]
```diff
@@ -264,11 +251,7 @@ impl<N: Network, E: Environment> Peer<N, E> {
                         trace!("Sending '{}-B' to {}", message.name(), peer_ip);
                         outbound_socket.send(message).await?;
 
-                        // Initialize a status variable.
-                        let status = Status::new();
-                        status.update(peer_status);
```
Comment (howardwu) [inline network/src/peer.rs:272]: Why was this removed during the process of a handshake?

File: network/src/peer.rs [hunk 2]
```diff
@@ -582,35 +560,35 @@ impl<N: Network, E: Environment> Peer<N, E> {
                                     // Update the node type of the peer.
                                     peer.node_type = node_type;
                                     // Update the status of the peer.
-                                    peer.status.update(status);
```
Comment (howardwu) [inline network/src/peer.rs:610]: If we can directly assign `status`, do we even need the `update()` method? Can/should it be removed?

File: network/src/prover.rs [hunk 3]
```diff
@@ -61,80 +60,55 @@ pub enum ProverRequest<N: Network> {
...
 /// A prover for a specific network on the node server.
 ///
-#[derive(Debug)]
 pub struct Prover<N: Network, E: Environment> {
     /// The state storage of the prover.
-    state: Arc<ProverState<N>>,
+    prover_state: Arc<ProverState<N>>,
     /// The Aleo address of the prover.
-    address: Option<Address<N>>,
+    pub address: Option<Address<N>>,
```
Comment (howardwu) [inline network/src/prover.rs]: revert: Why is this `pub` visibility? Can we move it into `State` if it needs to be accessed?

Also, why is `State` `pub` as well? Can we have `E: Environment` provide it?

File: network/src/prover.rs [hunk 4]
```diff
@@ -318,13 +328,15 @@ impl<N: Network, E: Environment> Prover<N, E> {
     ///
     /// Initialize the miner, if the node type is a miner.
     ///
-    async fn start_miner(prover: Arc<Self>, local_ip: SocketAddr) {
+    pub async fn start_miner(&self) {
```
Comment (howardwu) [inline network/src/prover.rs]: revert: Why is this `pub`? These methods are meant to be internal, otherwise there is "crossing of wires" which can lead to unintentional node halts

File: network/src/prover.rs [hunk 5]
```diff
@@ -203,13 +207,13 @@ impl<N: Network, E: Environment> Prover<N, E> {
     ///
     /// Sends a `PoolRegister` message to the pool IP address.
     ///
-    async fn send_pool_register(&self) {
+    pub async fn send_pool_register(&self) {
```
Comment (howardwu) [inline network/src/prover.rs]: revert: Why is this `pub`? These methods are meant to be internal, otherwise there is "crossing of wires" which can lead to unintentional node halts

File: network/src/ledger.rs [hunk 6]
```diff
@@ -195,13 +169,13 @@ impl<N: Network, E: Environment> Ledger<N, E> {
     /// Performs the given `request` to the ledger.
     /// All requests must go through this `update`, so that a unified view is preserved.
     ///
-    pub(super) async fn update(&self, request: LedgerRequest<N>) {
+    pub async fn update(&self, request: LedgerRequest<N>) {
```
Comment (howardwu) [inline network/src/ledger.rs]: revert: Why is `super` removed? The goal is not ensure this is not callable outside of this scope, out of order executions lead to node halts

File: network/src/operator.rs [hunk 7]
```diff
@@ -204,42 +161,40 @@ impl<N: Network, E: Environment> Operator<N, E> {
...
-        self.state.get_provers()
+        self.operator_state.get_provers()
     }
 
     ///
     /// Performs the given `request` to the operator.
     /// All requests must go through this `update`, so that a unified view is preserved.
     ///
-    pub(super) async fn update(&self, request: OperatorRequest<N>) {
+    pub async fn update(&self, request: OperatorRequest<N>) {
```
Comment (howardwu) [inline network/src/operator.rs]: Revert: why is `super` removed? The goal is not ensure this is not callable outside of this scope, out of order executions lead to node halts

File: network/src/state.rs [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+        Self {
+            local_ip,
+            peers: Default::default(),
+            ledger: Default::default(),
+            operator: Default::default(),
+            prover: Default::default(),
+        }
+    }
+
+    pub async fn initialize_peers(self: &Arc<Self>, peers: Peers<N, E>, mut peers_handler: PeersHandler<N, E>) {
```
Comment (howardwu) [inline network/src/state.rs:58]: remove: Why are these `initialize` methods `pub`?

**Comment by howardwu** [comment]
Overall, this structure is the way to go. Need to ensure the visibility does not leak, I've had some bad experiences with node halts when these methods are misused during future modifications

File: network/src/peer.rs (hunk 1 above)
Comment (ljedrz) [reply network/src/peer.rs:272]: I noticed this by accident - it's just redundant, as it creates an intermediate `status` binding that takes the value of `peer_status`, while we can just use `peer_status`

File: network/src/peer.rs (hunk 2 above)
Comment (ljedrz) [reply network/src/peer.rs:610]: with `Peer`, we have the option of mutating this field during synchronous access (so we can just assign to it directly), but for the immutable `Node` shared across multiple threads we still need to use `Status::update` that uses atomics

**Comment by ljedrz** [comment]
Rebased and applied some of the review suggestions related to APIs (some of the `pub`s seem to have been left over from an intermediate step of the refactoring).

File: network/src/prover.rs (hunk 3 above)
Comment (ljedrz) [reply network/src/prover.rs]: `state` doesn't need to be `pub` and it was removed; `address` is `pub` due to the `Operator` depending on it and there not being a method for it (as is the case with e.g. `MemoryPool`); I added a commit moving it to `State` like you suggested

File: network/src/state.rs (hunk 8 above)
Comment (ljedrz) [reply network/src/state.rs:58]: it's because they are called by the `Server`, and can also be called in tests that don't require the functionalities of the `Server` (i.e. just the state); it's very unlikely that those would be called by accident or more than once

**Comment by howardwu** [comment]
Have we been able to burn-in test this PR?

**Comment by jhaineault1** [comment]
TEST TEST TEST 
     Snarkos on $devnet-ue2c-arm-a1e568c376e78b13956d66e24f4b4c84 has been recompiled and docker containers have been rebuilt. 
     IP: $18.223.50.176 
     LOGS: $https://app.datadoghq.com/logs/livetail?query=pr%3A1727%20Report&cols=host%2Ccontainer_name

**Comment by jhaineault1** [comment]
Snarkos on devnet-ue2c-arm-a1e568c376e78b13956d66e24f4b4c84 has been recompiled and docker containers have been rebuilt. 
     IP: 18.223.50.176 
     LOGS: https://app.datadoghq.com/logs/livetail?query=pr%3A1727%20Report&cols=host%2Ccontainer_name

**Comment by jhaineault1** [comment]
Snarkos on devnet-ue2c-arm-a1e568c376e78b13956d66e24f4b4c84 has been recompiled and docker containers have been rebuilt. 
     LOGS: https://app.datadoghq.com/logs/livetail?query=pr%3A1727%20Report&cols=host%2Ccontainer_name

**Comment by jhaineault1** [comment]
Snarkos on devnet-ue2c-arm-a1e568c376e78b13956d66e24f4b4c84 has been recompiled and docker containers have been rebuilt. 
     LOGS: https://app.datadoghq.com/logs/livetail?query=pr%3A1727%20Report&cols=host%2Ccontainer_name

**Comment by zosorock** [comment]
> Have we been able to burn-in test this PR?

Yes, it has been running on an airtight devnet since last Thursday.

**Review by howardwu — APPROVED** [review]
